subsys/bluetooth/controller/ticker/ticker.c

```diff
@@ -329,93 +295,68 @@ static void ticker_by_next_slot_get(struct ticker_instance *instance,
+ * node in the linked node list.
  *
  * @param instance Pointer to ticker instance
  * @param id       Ticker node id to enqueue
  *
- * @return Id of enqueued ticker node, or id of previous- or colliding
- * ticker node if new node was not enqueued
+ * @return Id of enqueued ticker node
  * @internal
  */
 static u8_t ticker_enqueue(struct ticker_instance *instance, u8_t id)
 {
 	struct ticker_node *ticker_current;
 	struct ticker_node *ticker_new;
 	u32_t ticks_to_expire_current;
-	u8_t ticker_id_slot_previous;
-	u32_t ticks_slot_previous;
 	struct ticker_node *node;
 	u32_t ticks_to_expire;
 	u8_t previous;
 	u8_t current;
-	u8_t collide;
 
 	node = &instance->node[0];
 	ticker_new = &node[id];
 	ticks_to_expire = ticker_new->ticks_to_expire;
 
-	collide = ticker_id_slot_previous = TICKER_NULL;
 	current = instance->ticker_id_head;
 	previous = current;
-	ticks_slot_previous = instance->ticks_slot_previous;
 
 	/* Find insertion point for new ticker node and adjust ticks_to_expire
 	 * relative to insertion point
 	 */
-	while ((current != TICKER_NULL) &&
-	       (ticks_to_expire >
+	previous = TICKER_NULL;
+
+	while ((current != TICKER_NULL) && (ticks_to_expire >=
 		(ticks_to_expire_current =
-		 (ticker_current = &node[current])->ticks_to_expire))) {
+		(ticker_current = &node[current])->ticks_to_expire))) {
+
 		ticks_to_expire -= ticks_to_expire_current;
 
-		if (ticker_current->ticks_slot != 0U) {
-			ticks_slot_previous = ticker_current->ticks_slot;
-			ticker_id_slot_previous = current;
-		} else {
-			if (ticks_slot_previous > ticks_to_expire_current) {
-				ticks_slot_previous -= ticks_to_expire_current;
-			} else {
-				ticks_slot_previous = 0U;
-			}
+		/* Check for timeout in same tick - prioritize according to
+		 * latency
+		 */
+		if (ticks_to_expire == 0 && (ticker_new->lazy_current >
+					     ticker_current->lazy_current)) {
+			ticks_to_expire = ticker_current->ticks_to_expire;
+			break;
 		}
+
 		previous = current;
 		current = ticker_current->next;
 	}
 
-	/* Check for collision for new ticker node at insertion point */
-	collide = ticker_by_slot_get(&node[0], current,
-				     ticks_to_expire + ticker_new->ticks_slot);
-
-	if ((ticker_new->ticks_slot == 0U) ||
-	    ((ticks_slot_previous <= ticks_to_expire) &&
-	     (collide == TICKER_NULL))) {
-		/* New ticker node has no slot ticks or there is no collision -
-		 * link it in and adjust ticks_to_expire to relative value
-		 */
-		ticker_new->ticks_to_expire = ticks_to_expire;
-		ticker_new->next = current;
-
-		if (previous == current) {
-			instance->ticker_id_head = id;
-		} else {
-			node[previous].next = id;
-		}
+	/* Link in new ticker node and adjust ticks_to_expire to relative value
+	 */
+	ticker_new->ticks_to_expire = ticks_to_expire;
+	ticker_new->next = current;
 
-		if (current != TICKER_NULL) {
-			node[current].ticks_to_expire -= ticks_to_expire;
-		}
+	if (previous == TICKER_NULL) {
+		instance->ticker_id_head = id;
```

Comment:
Intent: Queue was empty (no previous), so set the new head.
However, it is possible that the very first loop through above, didn't set previous (because ticks_to_expire == 0 did a break before updating previous)

Reply:
ticks_to_expire cannot be 0